tests/test_spidermiddleware.py

```diff
@@ -0,0 +1,340 @@
+
+import logging
+
+from testfixtures import LogCapture
+from twisted.trial.unittest import TestCase
+from twisted.internet import defer
+
+from scrapy.spiders import Spider
+from scrapy.item import Item, Field
+from scrapy.http import Request
+from scrapy.utils.test import get_crawler
+
+
+class TestItem(Item):
+    value = Field()
+
+
+# ================================================================================
+# exceptions from a spider's parse method
+class BaseExceptionFromParseMethodSpider(Spider):
+    start_urls = ["http://example.com/"]
```

Comment:
It looks like these are real spiders making requests to http://example.com; that's not good to make http requests to external services in a testing suite; it is better to use scrapy's [mockserver](https://github.com/scrapy/scrapy/blob/master/tests/mockserver.py).